python-lib/dku_data_processing/format.py

```diff
@@ -31,6 +32,62 @@ def wkt_parser(wkt_point):
         return result[0]
 
 
+def extract_df(df, detail, filters, geopoint, tooltips):
+    """
+
+    Extract and select only necessary data to send to front-end.
+    Handle the filtering, normalizing, parsing and the tooltip.
+
+    :param detail:
+    :param df:
+    :param filters:
+    :param geopoint:
+    :param tooltips:
+    """
+    # Handle filtering
+    if (filters is not None) and (filters != []):
+        df = filter_dataframe(df, filters)
+    if df.empty:
+        return []
+
+    # Handle normalizing of detail column
+    if (detail is None) or (df.dtypes[detail] not in ['int64', 'float64']):
+        df['detail_'] = 1
+    else:
+        df['detail_'] = normalize(df[detail].values)
+    if (detail is not None) and (df.dtypes[detail] not in ['int64', 'float64']):
+        raise ValueError("The detail column should be a int or float type")
+
+    def convert(x):
+        if x is not None:
```

Comment:
Nitpicking - you can replace 'if x is not None` with `if x`(but maybe if x is not none is more readable, up to you)